robots/purePursuitController.h

```diff
@@ -0,0 +1,169 @@
+/*
+    Pure Pursuit Controller. The controller takes a list of waypoints as a path to follow,
+    then it calculates the optimal turning angle to stay on the path
+*/
+
+// BoB robotics includes
+#include "common/main.h"
+#include "robots/car_display.h"
+
+// Third-party includes
+#include "third_party/units.h"
+
+// Standard C++ includes
+#include <chrono>
+#include <iostream>
+#include <thread>
+
+// sign function
+template <typename T> int sgn(T val) {
+    return (T(0) < val) - (val < T(0));
+}
+
+namespace BoBRobotics {
+namespace Robots {
+
+using namespace units::math;
+using namespace units::literals;
+
+class PurePursuitController {
+
+    using millimeter_t = units::length::millimeter_t;
+    using degree_t = units::angle::degree_t;
+    using radian_t = units::angle::radian_t;
+   
+public:
+
+    PurePursuitController() {
+
+    }
+
+    PurePursuitController(millimeter_t lookahead, millimeter_t wheelBaseLength) : 
+                          lookAheadDistance(lookahead), wheelBase(wheelBaseLength) {
+
+    }
+
+    //! set waypoints which forms a path to be followed
+    void setWayPoints(std::vector<Vector2<millimeter_t>> wp) {
+        wayPoints = wp;
+    }
+
+    //! adds to the list of waypoints
+    void addToWayPoint(Vector2<millimeter_t> wayPoint) {
+        wayPoints.push_back(wayPoint);
+    }
+
+    //! sets the lookahead distance. Large values causes the car to cut corners,
+    //! where small values makes the car follow the path more closely
+    void setLookAheadDistance(millimeter_t distance) {
+        lookAheadDistance = distance;
+    }
+
+    //! sets the distance between the car's wheel bases
+    void setWheelBaseLength(millimeter_t length) {
+        wheelBase = length;
+    }
+  
+    //! calculates the turning angle needed to follow the path 
+    degree_t getTurningAngle(millimeter_t x, millimeter_t y, radian_t heading) {
+        std::vector<millimeter_t> lookPoint = getLookAheadPoint(x,y, lookAheadDistance);
+        if (!lookPoint.empty()) return computeTurningAngle(x,y, lookPoint.at(0), lookPoint.at(1), heading);
+        return 0_deg; 
+    }
+
+    //! calculates the look-ahead point the robot follows
+    std::vector<millimeter_t> getLookAheadPoint(millimeter_t x, millimeter_t y, millimeter_t r) {
```

Comment:
It would be nicer if this returned ``Vector2<millimeter_t> ``

Reply:
:+1: